### PR TITLE
Fix the potential hash code collison as an ID

### PR DESCRIPTION
**Pull request by qdongxu**

### CODE REVIEW

File: core/src/main/java/com/megaease/easeagent/core/Register.java [hunk 1]
```diff
@@ -36,15 +36,14 @@
...
+    private final Set<String> applied = new HashSet<>();
 
     Register(Iterable<QualifiedBean> beans) {
         this.beans = beans;
-        applied = new HashSet<Integer>();
     }
 
     void apply(String adviceClassName, ClassLoader external) {
-        if (!applied.add(Objects.hashCode(external, adviceClassName))) return;
+        if (!applied.add(adviceClassName + external.getClass().getName())) return;
```
Comment (akwei) [inline core/src/main/java/com/megaease/easeagent/core/Register.java:46]: This method has a problem:
```
adviceClassName = "111" , ClassLoader external = "112"  key="111112"
adviceClassName = "1111" , ClassLoader external = "12"  key="111112"

```
We should use a separator like `-`.


```
separator="-"
adviceClassName = "111" , ClassLoader external = "112"  key="111-112"
adviceClassName = "1111" , ClassLoader external = "12"  key="1111-12"

```

Comment (qdongxu) [reply core/src/main/java/com/megaease/easeagent/core/Register.java]: updated. and just found that it's incorrect to use the class name of the `external`.

File: core/src/main/java/com/megaease/easeagent/core/Register.java [hunk 2]
```diff
@@ -43,7 +44,7 @@
     }
 
     void apply(String adviceClassName, ClassLoader external) {
-        if (!applied.add(adviceClassName + external.getClass().getName())) return;
+        if (!applied.add(adviceClassName + "-" + System.identityHashCode(external))) return;
```
Comment (akwei) [inline core/src/main/java/com/megaease/easeagent/core/Register.java:47]: `Object.hashCode`、`Objects.hashCode `、`System.identityHashCode` cannot  guarantee the uniqueness of the object.
The object should be used as the key:
```
class ComboKey{
    String adviceClassName;
    ClassLoader classLoader;
}
```

File: core/src/main/java/com/megaease/easeagent/core/Register.java [hunk 3]
```diff
@@ -26,25 +26,34 @@
...
+    private final WeakHashMap<ClassLoader, Set<String>> applied = new WeakHashMap<>();
 
     Register(Iterable<QualifiedBean> beans) {
         this.beans = beans;
-        applied = new HashSet<Integer>();
     }
 
-    void apply(String adviceClassName, ClassLoader external) {
-        if (!applied.add(Objects.hashCode(external, adviceClassName))) return;
+    synchronized void apply(String adviceClassName, ClassLoader external) {
```
Comment (akwei) [inline core/src/main/java/com/megaease/easeagent/core/Register.java:47]: The instance of `adviceClassName` should be created once.
If adviceClassName = "a.b.UserAdvice":

1. `external = a`, Set<String> classNames = applied.get(external); ClassNames will be null, and classNames.add(adviceClassName) return true, then nwInstanceOf will be invoked.

2. `external = b`, Set<String> classNames = applied.get(external); ClassNames will be null, and classNames.add(adviceClassName) return true, then newInstanceOf will be invoked.

The instance of `adviceClassName` be created twice.

Comment (akwei) [inline core/src/main/java/com/megaease/easeagent/core/Register.java:47]: The instance of `adviceClassName` should be created once.
If adviceClassName = "a.b.UserAdvice":

1. `external = a`, Set<String> classNames = applied.get(external); ClassNames will be null, and classNames.add(adviceClassName) return true, then nwInstanceOf will be invoked.

2. `external = b`, Set<String> classNames = applied.get(external); ClassNames will be null, and classNames.add(adviceClassName) return true, then newInstanceOf will be invoked.

The instance of `adviceClassName` be created twice.

**Comment by qdongxu** [comment]
Resolved by #93

**Pull request closed without merging**